src/loader/riscv/pe/mod.rs

```diff
@@ -0,0 +1,253 @@
+    /// Unable to read Image header.
+    ReadImageHeader,
+    /// Unable to read DTB image
+    ReadDtbImage,
+    /// Invalid Image binary.
+    InvalidImage,
+    /// Invalid Image magic2 number.
+    InvalidImageMagicNumber,
+    /// Invalid base address alignment
+    InvalidBaseAddrAlignment,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let desc = match self {
+            Error::SeekImageEnd => "unable to seek Image end",
+            Error::SeekImageHeader => "unable to seek Image header",
+            Error::ReadImageHeader => "unable to read Image header",
+            Error::ReadDtbImage => "unable to read DTB image",
+            Error::SeekDtbStart => "unable to seek DTB start",
+            Error::SeekDtbEnd => "unable to seek DTB end",
+            Error::InvalidImage => "invalid Image",
+            Error::InvalidImageMagicNumber => "invalid Image magic2 number",
+            Error::DtbTooBig => "device tree image too big",
+            Error::ReadKernelImage => "unable to read kernel image",
+            Error::InvalidBaseAddrAlignment => "base address not aligned to 2MiB (for riscv64)",
+        };
+
+        write!(f, "PE Kernel Loader: {}", desc)
+    }
+}
+
+impl std::error::Error for Error {}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone, Default)]
+// See kernel doc Documentation/riscv/boot-image-header.rst
+// All these fields should be little endian.
+struct riscv_image_header {
+    code0: u32,
+    code1: u32,
+    text_offset: u64,
+    image_size: u64,
+    flags: u64,
+    version: u32,
+    res1: u32,
+    res2: u64,
+    magic: u64,
+    magic2: u32,
+    res3: u32,
+}
+
+impl KernelLoader for PE {
+    /// Loads a PE Image into guest memory.
+    ///
+    /// # Arguments
+    ///
+    /// * `guest_mem` - The guest memory where the kernel image is loaded.
+    /// * `kernel_offset` - 2MiB-aligned (for riscv64) base address in guest memory at which to load the kernel.
+    /// * `kernel_image` - Input Image format kernel image.
+    /// * `highmem_start_address` - ignored on RISC-V.
+    ///
+    /// # Returns
+    /// * KernelLoaderResult
+    fn load<F, M: GuestMemory>(
+        guest_mem: &M,
+        kernel_offset: Option<GuestAddress>,
+        kernel_image: &mut F,
+        _highmem_start_address: Option<GuestAddress>,
+    ) -> Result<KernelLoaderResult>
+    where
+        F: Read + Seek,
+    {
+        let kernel_size = kernel_image
+            .seek(SeekFrom::End(0))
+            .map_err(|_| Error::SeekImageEnd)? as usize;
+        let mut riscv_header: riscv_image_header = Default::default();
+        kernel_image.rewind().map_err(|_| Error::SeekImageHeader)?;
+
+        riscv_header
+            .as_bytes()
+            .read_from(0, kernel_image, mem::size_of::<riscv_image_header>())
+            .map_err(|_| Error::ReadImageHeader)?;
+
+        // Skip testing `magic` field as it's deprecated as of RISC-V boot image header version 0.2
```

Comment:
I don't think it's a good idea to have commented out code.

Reply:
Resolved by remove commented out code.